File: frontend/lib/models/property_model.dart
import 'dart:convert';

class Property {
  final int id;
  final String title;
  final String description;
  final String type;
  final double pricePerNight;
  final String location;
  final String city;
  final int bedrooms;
  final int beds;
  final int bathrooms;
  final int maxGuests;
  final List<String> amenities;
  final List<String> images;
  final double rating;
  final int? hostId;

  Property({
    required this.id,
    required this.title,
    required this.description,
    required this.type,
    required this.pricePerNight,
    required this.location,
    required this.city,
    required this.bedrooms,
    required this.beds,
    required this.bathrooms,
    required this.maxGuests,
    required this.amenities,
    required this.images,
    required this.rating,
    this.hostId,
  });

  factory Property.fromJson(Map<String, dynamic> json) {
    return Property(
      id: json['id'],
      title: json['title'] ?? 'No Title',
      description: json['description'] ?? '',
      type: json['type'] ?? 'Apartment',
      pricePerNight: double.tryParse(json['price_per_night']?.toString() ?? '0') ?? 0.0,
      location: json['location'] ?? 'Unknown',
      city: json['city'] ?? 'Unknown',
      bedrooms: json['bedrooms'] ?? 1,
      beds: json['beds'] ?? 1,
      bathrooms: json['bathrooms'] ?? 1,
      maxGuests: json['max_guests'] ?? 4,
      amenities: json['amenities'] is String 
          ? List<String>.from(jsonDecode(json['amenities'])) 
          : List<String>.from(json['amenities'] ?? []),
      images: json['images'] is String 
          ? List<String>.from(jsonDecode(json['images'])) 
          : List<String>.from(json['images'] ?? []),
      rating: double.tryParse(json['rating']?.toString() ?? '0') ?? 0.0,
      hostId: json['host_id'],
    );
  }
}

File: backend/models/propertyModel.js
const db = require('../config/db');

const Property = {
    getAll: async () => {
        const [rows] = await db.execute('SELECT * FROM properties');
        return rows;
    },

    getById: async (id) => {
        const [rows] = await db.execute('SELECT * FROM properties WHERE id = ?', [id]);
        return rows[0];
    },

    create: async (data, host_id) => {
        const { title, description, type, price_per_night, location, city, bedrooms, beds, bathrooms, amenities, images } = data;
        const [result] = await db.execute(
            'INSERT INTO properties (title, description, type, price_per_night, location, city, bedrooms, beds, bathrooms, amenities, host_id, images) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)',
            [title, description, type, price_per_night, location, city, bedrooms, beds, bathrooms, JSON.stringify(amenities), host_id, JSON.stringify(images)]
        );
        return result.insertId;
    },

    update: async (id, data) => {
        const { title, description, type, price_per_night, location, city, bedrooms, beds, bathrooms, amenities, images, rating } = data;
        await db.execute(
            'UPDATE properties SET title = ?, description = ?, type = ?, price_per_night = ?, location = ?, city = ?, bedrooms = ?, beds = ?, bathrooms = ?, amenities = ?, images = ?, rating = ? WHERE id = ?',
            [title, description, type, price_per_night, location, city, bedrooms, beds, bathrooms, JSON.stringify(amenities), JSON.stringify(images), rating, id]
        );
    },

    delete: async (id) => {
        await db.execute('DELETE FROM properties WHERE id = ?', [id]);
    }
};

module.exports = Property;
